server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -522,4 +528,22 @@ public SimilarityService getSimilarityService() {
     public BitsetFilterCache getBitsetFilterCache() {
         return bitsetFilterCache;
     }
+
+    public Set<String> sourcePaths(String name) {
+        return mapperSnapshot.sourcePaths(name);
+    }
+
+    /**
+     * Is the source stored?
+     */
+    public boolean isSourceStored() {
+        return mapperSnapshot.isSourceStored();
+    }
+
+    /**
+     * Returns the best nested {@link ObjectMapper} instances that is in the scope of the specified nested docId.
+     */
+    public ObjectMapper findNestedObjectMapper(int nestedDocId, IndexSearcher searcher, LeafReaderContext context) throws IOException {
```

Comment:
I think that if we make these changes incrementally we have the chance to clean this up a bit nicer, hopefully adding less methods to the different contexts. Lets see hoe far we get